src/utils/document-section-utils.ts

```diff
@@ -0,0 +1,267 @@
+  const mutatedStrucutre: SectionNodeStructure[] = structuredClone (structure);
+
+  for (const section of mutatedStrucutre) {
+    const baseSection = { ...section };
+    delete baseSection.children;
+    const flattenedSection: FlatSectionNodeStructure = {
+      ...baseSection,
+      childrenIds: [],
+    };
+
+    if (section.children.length) {
+      for (const child of section.children) {
+        flattenedSection.childrenIds.push(child.id);
+      }
+
+      flattenedStructure.push(flattenedSection);
+      flattenedStructure.push(...flattenStructure(section.children));
+    } else {
+      delete section.children;
+      flattenedStructure.push(flattenedSection);
+    }
+  }
+
+  return flattenedStructure;
+}
+
+function addSelectedSection(
+  flatStructure: FlatSectionNodeStructure[],
+  sectionId: string,
+  isChildId = false,
+): string[] {
+  const section = flatStructure.find(section => section.id === sectionId);
+  const sectionIdsToAdd = [sectionId];
+
+  if (isChildId) {
+    toggleDisableInput(sectionId, false);
+  }
+
+  if (section.childrenIds.length) {
+    toggleAllSectionChildElements(flatStructure, sectionId, false);
+    sectionIdsToAdd.push(...getCheckedChildSection(section, flatStructure));
+  }
+
+  for (const section of flatStructure) {
+    if (section.childrenIds.includes(sectionId)) {
+      section.isChecked = true;
+      sectionIdsToAdd.push(...addSelectedSection(flatStructure, section.id, true));
+    }
+  }
+
+  return sectionIdsToAdd;
+}
+
+function getCheckedChildSection(section: FlatSectionNodeStructure, flatStructure: FlatSectionNodeStructure[]): string [] {
+  const checkedChildSection: string[] = [];
+
+  for (const id of section.childrenIds) {
+    const childSection = flatStructure.find(section => section.id === id);
+
+    if (childSection?.isChecked) {
+      checkedChildSection.push(childSection.id);
+      checkedChildSection.push(...getCheckedChildSection(childSection, flatStructure));
+    }
+  }
+
+  return checkedChildSection;
+}
+
+export function toggleAllSectionChildElements(
+  flatStructure: FlatSectionNodeStructure[],
+  sectionId: string,
+  isDisabled: boolean
+): void {
+  const section = flatStructure.find(section => section.id === sectionId);
+
+  if (section?.childrenIds.length && section?.isChecked) {
+    for (const id of section.childrenIds) {
+      toggleDisableInput(id, isDisabled);
+      toggleAllSectionChildElements(flatStructure, id, isDisabled);
+    }
+  }
+}
+
+export function filterDocumentSections(
+  renderedDoc: HTMLElement,
+  nodes: Node[],
+  excludedNodes: string[],
+  storedStyles: StoredStyle[]
+): HTMLElement {
+  const proseMirrorContainer = renderedDoc.getElementsByClassName('ProseMirror')[0] ?? null;
+
+  if (proseMirrorContainer) {
+    const tempTocNodeList = JSON.parse(JSON.stringify(nodes));
+
+    for (const id of excludedNodes) {
+      const node = nodes.find(node => node.attrs.objectId === id) ?? null;

```

Comment:
is the `?? null` necessary here? I don't see any null checks so undefined should be fine. (?? counts as a branch)